rpc/rpc_analytics_purger.go

```diff
@@ -119,35 +117,29 @@ func (r *Purger) PurgeCache() {
 			//if it's the first iteration, we look for tyk-system-analytics to maintain backwards compatibility or if analytics_config.enable_multiple_analytics_keys is disabled in the gateway
 			analyticsKeyName = ANALYTICS_KEYNAME
 		} else {
+			// keyname + serializationmethod
 			analyticsKeyName = fmt.Sprintf("%v_%v", ANALYTICS_KEYNAME, i)
 		}
 
-		analyticsValues := r.Store.GetAndDeleteSet(analyticsKeyName)
-		if len(analyticsValues) == 0 {
-			continue
-		}
-		keys := make([]interface{}, len(analyticsValues))
-
-		for i, v := range analyticsValues {
-			decoded := AnalyticsRecord{}
-			if err := msgpack.Unmarshal([]byte(v.(string)), &decoded); err != nil {
-				Log.WithError(err).Error("Couldn't unmarshal analytics data")
-			} else {
-				Log.WithField("decoded", decoded).Debug("Decoded Record")
-				keys[i] = decoded
+		for _, serializerMethod := range AnalyticsSerializers {
+
+			analyticsKeyName += serializerMethod.GetSuffix()
+			analyticsValues := r.Store.GetAndDeleteSet(analyticsKeyName)
+			if len(analyticsValues) == 0 {
+				continue
 			}
-		}
 
-		data, err := json.Marshal(keys)
-		if err != nil {
-			Log.WithError(err).Error("Failed to marshal analytics data")
-			return
-		}
+			keys := make([][]byte, len(analyticsValues))
+			for i, v := range analyticsValues {
+				keys[i] = []byte(v.(string))
+			}
 
-		// Send keys to RPC
-		if _, err := FuncClientSingleton("PurgeAnalyticsData", string(data)); err != nil {
-			EmitErrorEvent(FuncClientSingletonCall, "PurgeAnalyticsData", err)
-			Log.Warn("Failed to call purge, retrying: ", err)
+			// Send keys as [][]bytes to RPC
```

Comment:
If we change from `string` to `[][]byte` we would lose backward compatibility. Imagine the scenario a user upgrades the gateway (without touching anything related to serialization type), everything should keep working.

Reply:
yes, I see what you mean. In tyk-sink side now it's defined as `Dispatch.AddFunc("PurgeAnalyticsData", func(clientID string, data [][]byte) error` . If we want to send multiple items at the same time then it need to change, otherwise we will need to loop here and send record by record to keep backward compatibility, send the string and the convert to bytes again inside `PurgeAnalyticsData`

Reply:
I think it's gonna be a trade-off with performance but it's totally necessary IMO

Reply:
this was reverted to keep sending strings